dips/dip-158.md

```diff
@@ -0,0 +1,805 @@
+
+## AddressObject
+Represents a physical address
+| Field | Type | Required? | Description |
+|-|-|-|-|
+| city | str | Y | The city, district, suburb, town, or village |
+| country | str | Y | Two-letter country code (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) |
+| line1 | str | Y | Address line 1 |
+| line2 | str | N | Address line 2 - apartment, unit, etc.|
+| postal_code| str | Y | ZIP or postal code |
+| state | str | N | State, county, province, region.|
+```
+{
+    "city": "San Francisco",
+    "country": "US",
+    "line1": "1260 Market Street",
+    "line2": "Suite 450",
+    "postal_code": "94103",
+    "state": "CA",
+}
+```
+## NationalIdObject
+Represents a national ID.
+
+| Field | Type | Required? | Description |
+|-|-|-|-|
+| id_value | str | Y | Indicates the national ID value - for example, a social security number |
+| country | str | N | Two-letter country code (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) |
+| type | str | N | Indicates the type of the ID |
+
+```
+{
+    "id_value": "123-45-6789",
+    "country": "US",
+    "type": "SSN",
+}
+```
+
+---
+# Appendix A - Prerequisite: Sharing Common Payment Identifier and Address
+
+Note that this is a prerequisite for the off-chain negotiations and not a part of it.
+
+## Acquire Payment Identifier and Receiver Address Using Redirect URL
+
+This URL contains the payment reference id (i.e. the common payment identifier) and the wallet's website address.
+
+The redirect URL is likely to be used in the **Web Redirect** flow which means that the customer will be redirected from the merchant's checkout page to the wallet's website to review and approve the payment. 
+
+The table below specifies the fields that should be encoded into a series of URL parameters appended to the query string.
+
+|Field|Type|Required?|Description|
+|-|-|-|-|
+|vasp_address|str|Y|Address of receiving VASP. The address is encoded using bech32. For Diem addresses format, refer to the "account identifiers" section in [DIP-5](https://dip.diem.com/dip-5/#account-identifiers).|
+|reference_id|str|Y|A unique identifier of this payment. It should be a UUID according to RFC4122 with "-"'s included.|
+|redirect_url|str|N|URL used by the wallet to redirect the customer back to the merchant|
+
+For example, for the following values:
+|Field|Value|
+|-|-|
+|vasp_address|dm1pgxah7pxhzljvp3p4m9g0m3tm0qqqqqqqqqqqqqqgyftgh| 
+|reference_id|ad8d888a-1791-4b63-98e5-6f1d6ddb4411|
+|redirect_url|somemerchantnname.com/callbackpage|
+
+The URL format would be:
+
+`https://some-diem-wallet.com/pay?vasp_address=dm1pgxah7pxhzljvp3p4m9g0m3tm0qqqqqqqqqqqqqqgyftgh&reference_id=ad8d888a-1791-4b63-98e5-6f1d6ddb4411&redirect_url=somemerchantnname.com/callbackpage`
+
+
+After parsing this request, the wallet can request the payment details from the receiving VASP and display them to the customer to review and approve. 
+
+## Acquire Payment Identifier and Receiver Address Using QR Code
```

Comment:
Should we consider leaving an optional QR code or is that a bit too quirky :P

Reply:
Did not understand this comment. Would appreciate some clarification.

Reply:
Done